File: Algorithms/Backtracking/python/combination_sum.py
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        """Basic problem to use the Combination template

        Args:
            candidates: Given a set of distinct integers.
            target:  A interger to meet the requirement.

        Returns:
            return all possible solutions with sum == target.

        Example:
            candidates = [2,3,6,7], target = 7

            Thus, res = [ [7], [2,2,3] ] 

        Time Complexity: O(n * 2^n)
        Space Complexity: O(2^n) 
        """
        res = []
        self.helper(res, [], candidates, target, 0, 0)
        return res

        return res

    def helper(self, res: List[int], cur: List[int], candidates: List[int], target: int, start: int, cur_sum: int) -> None:
        """Main backtracking function

        Args:
            res: A List to save the all possible solutions.
            cur: A List to save the current solution.
            candidates: The same input. i.e. Given a set of distinct integers.
            target: A interger to meet the requirement.
            start: the index start point for exploring.
            cur_sum: the sum of cur array.


        Returns:
            None

        * Note: the performance can be slightly better if we choose the remaining = tartget - arr[i] (i.e. O(1)),
        intead of sum all the list (i.e. O(n)). However, consider the time complexity of backtracking is larger than O(n),
        it is relative trivial to optimize it here. 

        """
        if cur_sum == target:
            res.append(cur[:])

        for i in range(start, len(candidates)):
            if cur_sum < target:
                cur.append(candidates[i])
                cur_sum += candidates[i]
                self.helper(res, cur, candidates, target, i, cur_sum)
                cur_sum -= cur.pop()

    # TODO: This question has a DP solution, will do later.
    def dp(self):
        """Dynamic Programming solution
        """
        return None

File: Algorithms/Backtracking/java/CombinationSum.java
import java.util.*;

/**
 * Given a set of candidate numbers (candidates) (without duplicates) and a
 * target number (target), find all unique combinations in candidates where the
 * candidate numbers sums to target. The same repeated number may be chosen from
 * candidates unlimited number of times.
 * 
 * Time Complexity: O(n * 2^n), Space Complexity: O(2^n)
 */
public class CombinationSum {
    /**
     * @param candidates: Given a set of distinct integers.
     * @param target:     A interger to meet the requirement.
     * @return all possible solutions with sum == target.
     */
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        helper(res, new ArrayList<Integer>(), candidates, target, 0, 0);
        return res;
    }

    /**
     * @param res:        A List to save the all possible solutions.
     * @param cur:        A List to save the current solution.
     * @param candidates: The same input. i.e. Given a set of distinct integers.
     * @param target:     A interger to meet the requirement.
     * @param start:      the index start point for exploring.
     * @param sum:        the sum of cur array.
     */
    private void helper(List<List<Integer>> res, List<Integer> cur, int[] candidates, int target, int start, int sum) {
        if (sum == target)
            res.add(new ArrayList(cur));
        for (int i = start; i < candidates.length; i++) {
            if (sum < target) {
                sum += candidates[i];
                cur.add(candidates[i]);
                helper(res, cur, candidates, target, i, sum);
                sum -= cur.get(cur.size() - 1);
                cur.remove(cur.size() - 1);
            }
        }
    }

    // TODO: Add a DP solution
    public List<List<Integer>> dp(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        return res;
    }
}